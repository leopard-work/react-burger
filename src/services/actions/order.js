import {createOrder} from "../../utils/api";

export const CHECKOUT_REQUEST = 'CHECKOUT_REQUEST';
export const CHECKOUT_SUCCESS = 'CHECKOUT_SUCCESS';
export const CHECKOUT_FAILED = 'CHECKOUT_FAILED';
export const CLEAR_ORDER = 'CLEAR_ORDER';

export function checkOutSend(body) {                                // Отправка заказа
    return function(dispatch) {
        dispatch({type: CHECKOUT_REQUEST});
        createOrder(body).then((data) => {
            dispatch({type: CHECKOUT_SUCCESS, order: data});
        }).catch(() => {
            dispatch({type: CHECKOUT_FAILED});
            setTimeout(() => {
                dispatch({type: CLEAR_ORDER});
            }, 3000);
        })
    }
}