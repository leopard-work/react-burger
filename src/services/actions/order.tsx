import { createOrder } from "../../utils/api";
import { BASKET_CLEAR } from "./basket";

export const CHECKOUT_REQUEST = "CHECKOUT_REQUEST";
export const CHECKOUT_SUCCESS = "CHECKOUT_SUCCESS";
export const CHECKOUT_FAILED = "CHECKOUT_FAILED";
export const CLEAR_ORDER = "CLEAR_ORDER";

export function checkOutSend(
  body: { ingredients: Array<string> },
  token: string
) {
  // Отправка заказа
  return function (dispatch: any) {
    dispatch({ type: CHECKOUT_REQUEST });
    createOrder(body, token)
      .then((data) => {
        dispatch({ type: CHECKOUT_SUCCESS, order: data });
        dispatch({ type: BASKET_CLEAR });
      })
      .catch(() => {
        dispatch({ type: CHECKOUT_FAILED });
        setTimeout(() => {
          dispatch({ type: CLEAR_ORDER });
        }, 3000);
      });
  };
}
