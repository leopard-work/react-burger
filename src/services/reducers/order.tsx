import {
  CHECKOUT_ACTIONS,
  CHECKOUT_FAILED,
  CHECKOUT_REQUEST,
  CHECKOUT_SUCCESS,
  CLEAR_ORDER,
} from "../actions/order";
import { OrderProps } from "../../utils/types";

type orderState = {
  orderInfo: {
    order: OrderProps;
  };
  orderModalOpen: boolean;
  orderRequest: boolean;
  orderFailed: boolean;
};

const initialState: orderState = {
  orderInfo: {
    order: {
      ingredients: [],
      _id: "",
      owner: {
        name: "",
        email: "",
        createdAt: "",
        updatedAt: "",
      },
      status: "",
      name: "",
      createdAt: "",
      updatedAt: "",
      number: null,
      price: 0,
    },
  }, // Информация о заказе
  orderModalOpen: false, // Открытие окна с заказом
  orderRequest: false,
  orderFailed: false,
};

export const orderReducer = (
  state = initialState,
  action: CHECKOUT_ACTIONS
) => {
  switch (action.type) {
    case CHECKOUT_REQUEST: {
      // Отправка заказа
      return {
        ...state,
        orderRequest: true,
      };
    }
    case CHECKOUT_SUCCESS: {
      return {
        ...state,
        orderRequest: false,
        orderFailed: false,
        orderModalOpen: true,
        orderInfo: action.order,
      };
    }
    case CHECKOUT_FAILED: {
      return {
        ...state,
        orderFailed: true,
        orderRequest: false,
      };
    }
    case CLEAR_ORDER: {
      // Удаление информации о заказе
      return {
        ...state,
        orderInfo: initialState.orderInfo,
        orderFailed: false,
        orderModalOpen: false,
      };
    }
    default: {
      return state;
    }
  }
};
