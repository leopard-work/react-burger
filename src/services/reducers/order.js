import {CHECKOUT_FAILED, CHECKOUT_REQUEST, CHECKOUT_SUCCESS, CLEAR_ORDER} from "../actions/order";

const initialState = {
    orderInfo: {},                                                                          // Информация о заказе
    orderModalOpen: false,                                                                  // Открытие окна с заказом
    orderRequest: false,
    orderFailed: false
};

export const orderReducer = (state = initialState, action) => {
    switch (action.type) {
        case CHECKOUT_REQUEST: {                                                            // Отправка заказа
            return {
                ...state,
                orderRequest: true
            }
        }
        case CHECKOUT_SUCCESS: {
            return {
                ...state,
                orderRequest: false,
                orderFailed: false,
                orderModalOpen: true,
                orderInfo: action.order,
            };
        }
        case CHECKOUT_FAILED: {
            return {
                ...state,
                orderFailed: true,
                orderRequest: false
            }
        }
        case CLEAR_ORDER: {                                                                 // Удаление информации о заказе
            return {
                ...state,
                orderInfo: {},
                orderFailed: false,
                orderModalOpen: false
            }
        }
        default: {
            return state;
        }
    }
}