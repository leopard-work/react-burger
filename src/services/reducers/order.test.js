import {
  CHECKOUT_ACTIONS,
  CHECKOUT_FAILED,
  CHECKOUT_REQUEST,
  CHECKOUT_SUCCESS,
  CLEAR_ORDER,
} from "../actions/order";

const testOrder = {
  ingredients: [],
  _id: "",
  owner: {
    name: "",
    email: "",
    createdAt: "",
    updatedAt: "",
  },
  status: "",
  name: "",
  createdAt: "",
  updatedAt: "",
  number: null,
  price: 0,
};

import { initialState, orderReducer } from "./order";
import renderer from "react-test-renderer";

import React from "react";

describe("Оформление заказа", () => {
  it("- исходное состояние", () => {
    expect(orderReducer(initialState, {})).toEqual(initialState);
    const tree = renderer.create(<>Исходное состояние</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка CHECKOUT_REQUEST", () => {
    expect(
      orderReducer(initialState, {
        type: CHECKOUT_REQUEST,
      })
    ).toEqual({
      ...initialState,
      orderRequest: true,
    });
    const tree = renderer.create(<>Запрос на отправку заказа</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка CHECKOUT_SUCCESS", () => {
    expect(
      orderReducer(initialState, {
        type: CHECKOUT_SUCCESS,
        order: testOrder,
      })
    ).toEqual({
      ...initialState,
      orderRequest: false,
      orderFailed: false,
      orderModalOpen: true,
      orderInfo: testOrder,
    });
    const tree = renderer.create(<>Отправка заказа</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка CHECKOUT_FAILED", () => {
    expect(
      orderReducer(initialState, {
        type: CHECKOUT_FAILED,
      })
    ).toEqual({
      ...initialState,
      orderFailed: true,
      orderRequest: false,
    });
    const tree = renderer.create(<>Ошибка отправки заказа</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
