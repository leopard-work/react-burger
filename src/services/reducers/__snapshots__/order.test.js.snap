// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Оформление заказа - исходное состояние 1`] = `"Исходное состояние"`;

exports[`Оформление заказа - обработка CHECKOUT_FAILED 1`] = `"Ошибка отправки заказа"`;

exports[`Оформление заказа - обработка CHECKOUT_REQUEST 1`] = `"Запрос на отправку заказа"`;

exports[`Оформление заказа - обработка CHECKOUT_SUCCESS 1`] = `"Отправка заказа"`;
